downstairs/src/region.rs

```diff
@@ -841,13 +842,54 @@ impl Region {
         Ok(())
     }
 
-    #[instrument]
+    /// Reads a region, returning a `Vec<ReadResponse>`
+    ///
+    /// (instead of a pre-serialized `Bytes` object)
+    #[cfg(test)]
     pub async fn region_read(
         &mut self,
         requests: &[crucible_protocol::ReadRequest],
         job_id: JobId,
-    ) -> Result<Vec<crucible_protocol::ReadResponse>, CrucibleError> {
-        let mut responses = Vec::with_capacity(requests.len());
+    ) -> Result<Vec<ReadResponse>, CrucibleError> {
+        let out = self.region_read_raw(requests, job_id).await?;
+        bincode::deserialize(&out).unwrap()
+    }
+
+    /// Reads data from a region
+    ///
+    /// Returns either an error or the bincode-serialized representation of a
+    /// `Result<Vec<ReadResponse>, CrucibleError>`, which will always be
+    /// `Ok(..)`.
+    #[instrument]
+    pub async fn region_read_raw(
+        &mut self,
+        requests: &[crucible_protocol::ReadRequest],
+        job_id: JobId,
+    ) -> Result<Bytes, CrucibleError> {
+        // We want to avoid an extra memcpy when serializing into the socket,
+        // which means we get to pre-serialize our data into a form that matches
+        // that of bincode.  By hand.  Sorry.
+
+        // We'll estimate capacity here, assuming 0 or 1 context per read.  This
+        // always true for the raw file extents; the SQLite extents may have to
+        // reallocate, but they're not optimized for speed anyways.
+        let capacity = core::mem::size_of::<u32>() // Ok(..)
+            + core::mem::size_of::<usize>() // Vec(..)
+            + requests.len()
+                * (core::mem::size_of::<u64>() // eid
+                    + core::mem::size_of::<u64>()  // block.value
+                    + core::mem::size_of::<u32>()  // block.shift
+                    + core::mem::size_of::<usize>()  // data size
+                    + self.def.block_size() as usize // read data
+                    + core::mem::size_of::<usize>() // context count
+                    + core::mem::size_of::<u64>() // hash
+                    + core::mem::size_of::<u8>() // option tag
+                    + core::mem::size_of::<[u8; 12]>() // nonce
+                    + core::mem::size_of::<[u8; 16]>()); // tag
+
```

Comment:
Is the above what exactly the data looks like with bincode?
If so, and in some future bincode changes, would this fail to compile?
Or is bincode changing it's format not a realistic scenario?

Reply:
Yes, this is exactly how the data is serialized.

`bincode`'s binary format is [specified](https://github.com/bincode-org/bincode/blob/trunk/docs/spec.md) and should not change (although I did open a ticket to [clarify `Option`'s behavior](https://github.com/bincode-org/bincode/issues/700))

I moved this blob into a standalone function and added a unit test in 8f04737bc10f4cb7036b8f318b0a84f89a1bb7c3

Reply:
>  moved this blob into a standalone function and added a unit test in

I'm I little disappointed you (correctly) branch predicted my follow on comment, robbing
me the chance to do so, but don't worry, I'll recover.